api/python/notebooks/comp_bio_normalizing_full_gene_sequencing.ipynb

```diff
@@ -0,0 +1,393 @@
+   "outputs": [],
+   "source": [
+    "sc.pp.calculate_qc_metrics(liver_adata, percent_top=None, log1p=False, inplace=True)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Normalizing expression to account for gene length\n",
+    "\n",
+    "By default `cell_census.get_anndata()` fetches all genes in the Cell Census. So let's first identify the genes that were measured in this dataset and subset the `AnnData` to only include those. \n",
+    "\n",
+    "We would normally do this by using the \"Dataset Presence Matrix\" in `census[\"census_data\"][\"mus_musculus\"].ms[\"RNA\"].varp[\"dataset_presence_matrix\"]`. But for simplicity let's just remove those genes that have 0 values across all cells."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "52373"
+      ]
+     },
+     "execution_count": 5,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "liver_adata.n_vars"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "gene_presence = [False] * liver_adata.X.shape[1]\n",
+    "for i in range(liver_adata.X.shape[1]):\n",
+    "    if liver_adata.X[:, i].count_nonzero() > 0:\n",
+    "        gene_presence[i] = True\n",
+    "\n",
+    "liver_adata = liver_adata[:, gene_presence].copy()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "16906"
+      ]
+     },
+     "execution_count": 7,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "liver_adata.n_vars"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We can see that out of the 52,373 genes in the Cell Census **16,906** were measured in this dataset. \n",
+    "\n",
+    "Now let's normalize these genes by gene length. We can easily do this because the Cell Census has gene lengths included in the gene metadata under `feature_length`."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "array([[   0.,    0.,    0.,    0.,    0.],\n",
+       "       [ 297.,  106.,    0.,    0.,    0.],\n",
+       "       [   0.,    0.,    0.,    0.,    0.],\n",
+       "       [2347.,  304.,    0.,  815.,   26.],\n",
+       "       [   0.,  787.,    0.,    0.,    0.]], dtype=float32)"
+      ]
+     },
+     "execution_count": 8,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "liver_adata.X[:5, :5].toarray()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "for i in range(liver_adata.X.shape[1]):\n",
```

Comment:
This is an anti-pattern - you should use numpy broadcasting to do the division.  By way of example, to divide across axis 1:
```
In [10]: (np.array([[1,2,3],[4,5,6]], dtype=np.float32).T / np.array([2,3])).T
Out[10]: 
array([[0.5       , 1.        , 1.5       ],
       [1.33333333, 1.66666667, 2.        ]])```

In your code, you would do something like:
```
liver_adata.X = (liver_adata.X.T / liver_adata.var.feature_length).T
```

CAVEAT: I did not test the above with the sparse API, so try it first :-)

Reply:
I had to do some casting but it works